File: src/composables/commands/web-storage.ts
import { invoke } from '@tauri-apps/api'

export const webStorageExport = (serviceId: string, packageId: string, path: string): Promise<void> => {
  return invoke('webStorageExport', { service: serviceId, package: packageId, path })
    .then((value) => {
      if (value)
        throw new TypeError(`Internal error: \`closeFileChunked\` has returned \`${value}\` but \`undefined | null\` was expected`)
    })
}

export const webStorageImport = (serviceId: string, packageId: string, path: string): Promise<void> => {
  return invoke('webStorageImport', { service: serviceId, package: packageId, path })
    .then((value) => {
      if (value)
        throw new TypeError(`Internal error: \`closeFileChunked\` has returned \`${value}\` but \`undefined | null\` was expected`)
    })
}

export const webStorageClear = (serviceId: string, packageId: string): Promise<void> => {
  return invoke('webStorageClear', { service: serviceId, package: packageId })
    .then((value) => {
      if (value)
        throw new TypeError(`Internal error: \`closeFileChunked\` has returned \`${value}\` but \`undefined | null\` was expected`)
    })
}

export const webStorageDelete = (serviceId: string, packageId: string, lectionId: string | undefined, key: string): Promise<string | undefined> => {
  return invoke('webStorageDelete', { service: serviceId, package: packageId, lection: lectionId, key })
    .then((value) => {
      if (typeof value === 'string')
        return value

      if (!value)
        return undefined

      throw new TypeError(`Internal error: \`webStorageDelete\` has returned \`${value}\` but \`string\` was expected`)
    })
}

export const webStorageGet = (serviceId: string, packageId: string, lectionId: string | undefined, key: string, fallback: string | undefined): Promise<string | undefined> => {
  return invoke('webStorageGet', { service: serviceId, package: packageId, lection: lectionId, key, fallback })
    .then((value) => {
      if (typeof value === 'string')
        return value

      if (!value)
        return undefined

      throw new TypeError(`Internal error: \`webStorageGet\` has returned \`${value}\` but \`string\` was expected`)
    })
}

export const webStorageSet = (serviceId: string, packageId: string, lectionId: string | undefined, key: string, value: string, replaced: boolean): Promise<string | undefined> => {
  return invoke('webStorageSet', { service: serviceId, package: packageId, lection: lectionId, key, value, replaced })
    .then((value) => {
      if (typeof value === 'string')
        return value

      if (!value)
        return undefined

      throw new TypeError(`Internal error: \`webStorageSet\` has returned \`${value}\` but \`string\` was expected`)
    })
}

File: src-tauri/src/commands/web_storage.rs
use crate::managers::ServiceManager;


type StateServiceManager<'l> = tauri::State<'l, std::sync::Arc<ServiceManager>>;


#[tauri::command]
#[allow(non_snake_case)]
pub async fn webStorageExport(services: StateServiceManager<'_>, service: &str, package: &str, path: &str) -> Result<(), String> {
    println!("export! {path}");
    services
        .access(service)
        .await
        .as_web_storage()
        .map_err(|error| error.to_string())?
        .borrow()
        .export(package, path)
        .await
        .map_err(|error| error.to_string())
}


#[tauri::command]
#[allow(non_snake_case)]
pub async fn webStorageImport(services: StateServiceManager<'_>, service: &str, package: &str, path: &str) -> Result<(), String> {
    println!("import! {path}");
    services
        .access(service)
        .await
        .as_web_storage()
        .map_err(|error| error.to_string())?
        .borrow()
        .import(package, path)
        .await
        .map_err(|error| error.to_string())
}


#[tauri::command]
#[allow(non_snake_case)]
pub async fn webStorageClear(services: StateServiceManager<'_>, service: &str, package: &str) -> Result<(), String> {
    services
        .access(service)
        .await
        .as_web_storage()
        .map_err(|error| error.to_string())?
        .borrow()
        .clear(package)
        .await
        .map_err(|error| error.to_string())
}


#[tauri::command]
#[allow(non_snake_case)]
pub async fn webStorageDelete(
    services: StateServiceManager<'_>,
    service: &str,
    package: &str,
    lection: Option<&str>,
    key: &str
) -> Result<Option<String>, String> {

    services
        .access(service)
        .await
        .as_web_storage()
        .map_err(|error| error.to_string())?
        .borrow()
        .delete(package, lection, key)
        .await
        .map_err(|error| error.to_string())
}


#[tauri::command]
#[allow(non_snake_case)]
pub async fn webStorageGet(
    services: StateServiceManager<'_>,
    service: &str,
    package: &str,
    lection: Option<&str>,
    key: &str,
    fallback: Option<&str>
) -> Result<Option<String>, String> {

    services
        .access(service)
        .await
        .as_web_storage()
        .map_err(|error| error.to_string())?
        .borrow()
        .get(package, lection, key, fallback)
        .await
        .map_err(|error| error.to_string())
}


#[tauri::command]
#[allow(non_snake_case)]
pub async fn webStorageSet(
    services: StateServiceManager<'_>,
    service: &str,
    package: &str,
    lection: Option<&str>,
    key: &str,
    value: &str,
    replaced: bool
) -> Result<Option<String>, String> {
    
    services
        .access(service)
        .await
        .as_web_storage()
        .map_err(|error| error.to_string())?
        .borrow()
        .set(package, lection, key, value, replaced)
        .await
        .map_err(|error| error.to_string())
}
